### PR TITLE
Only check fls/dls license if the feature is used

### PR DESCRIPTION
**Pull request by rjernst**
This commit moves the check of licensing for FLS/DLS to later in the
reader wrapper, so that if no permissions exist, the license is not
checked. This delay then ensures the license is only checked (and
logged) when the feature is actually used.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/SecurityIndexReaderWrapper.java [hunk 1]
```diff
@@ -81,6 +80,11 @@ public DirectoryReader apply(final DirectoryReader reader) {
                 return reader;
             }
 
+            // permissions are defined, so check if the license actually allows DLS/FLS
+            if (licenseState.checkFeature(Feature.SECURITY_DLS_FLS) == false) {
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/SecurityIndexReaderWrapper.java:84]: At this point in the code we have actually established that the current role applies DLS or FLS, only that there is an `IndicesAccessControl` object.
An `IndicesAccessControl` object in the thread context only means that we performed access checks at some point, it doesn't mean DLS is enabled.

For that we need to check  `permissions.getDocumentPermissions().hasDocumentLevelPermissions()`

**Review by tvernum — COMMENTED** [review]
I think this PR is evidence that our approach to feature usage checking is inherently fragile.

All it takes is for some code to call `checkFeature` before being 100% confident that the feature is actually going to be used and we end up with incorrect reporting.

Even if this PR is updated to call `checkFeature` in the right place, how do ensure that some future refactoring doesn't move it around again?
We would need tests that almost-but-not-quite use a feature and then check that the usage tracking didn't report it.

I'm starting to question whether explicit "feature has been used" calls wouldn't be better than transparently assuming that every check for a feature must imply that it's in use.

**Comment by rjernst** [comment]
@tvernum Your points are fair. But the fragility you mention is also the advantage of this approach, and exactly why we went with it to start. While it is possible a feature is marked as used at the incorrect time (to frequently or not when it is actually used), it ensures that we don't *miss* any use cases entirely. With a completely separate function that must be called for tracking we could easily check the license, but then fail to track the feature's usage. IMO the current approach is inherently better for that reason. The tracking can be worked out, but we won't ever not track a licensed feature.

As far as ensuring we call checkFeature in the right place, I think that is up to each individual set of licensed code, since each has its own conditions that must be mimicked to ensure the tracking is and is not done. Note that we would need the same tests regardless of whether we had a separate tracker or not.

**Comment by tvernum** [comment]
> it ensures that we don't miss any use cases entirely

There are currently 19 features that are Gold+
We will increase that slightly by splitting up the realms, but only slightly, and we will add new features but that's harder to quantify.

If we exclude features that aren't available on cloud (IP Filtering, Authz Engines, AD, LDAP) and features that we don't bill for (Operator Privileges, Token Service), we end up with less than 20 features that we need to track.

It does not feel like an insurmountable task to have integration tests for each of those features that ensure that the feature is tracked when under the conditions that trigger its "use".

The alternative problem, of how do we make sure we're not inadvertently marking a feature used when it actually wasn't, is harder to guard against. We could have an integTest that uses a lot of basic/standard licensed functionality and then verifies that the usage API hasn't reported anything Gold+, but it's a bit of a weird approach and hard to be confident about.

Part of the issue here is that `checkFeature` has a side effect - it records the feature as used. That side effect is not obvious from its name, and unless you have that special knowledge at front of brain, you cannot infer from reading a piece of code that it will trigger feature usage, and whether it is triggered in the right place.
I'm worried that we're going to be have recurring issues where small refactoring we treat a feature as used when it wasn't and that actually seems like a bigger problem than missing something.

**Comment by rjernst** [comment]
> Part of the issue here is that checkFeature has a side effect

That's a fair criticism. However, I still contend that the side effect is the advantage; it ensures we haven't skipped tracking for a feature because we are tracking the license checking. I originally named this just checkFeature before we were tracking. Since all features will soon be tracked (when BASIC is gone from the Feature enum), would a change to the name like `checkAndTrackFeature` ease some of your concerns?

**Comment by rjernst** [comment]
Decided to go a different way with this, so closing.

**Pull request closed without merging**